### PR TITLE
feat: exclude doctors with a current F1 placement from the reval sync…

### PR DESCRIPTION
**Pull request by CaiWillis**
… and CDC

TIS21-4459

### CODE REVIEW

File: tcs-persistence/src/main/resources/queries/traineeInfoForConnection.sql [hunk 1]
```diff
@@ -46,8 +47,18 @@ from (
...
     group by cm2.programmeMembershipUuid
   ) latestCm on latestCm.programmeMembershipUuid = pm1.uuid
+  left join (
+      select
+  		pl.traineeId,
+  		-- one row per trainee
+  		GROUP_CONCAT(pl.gradeAbbreviation SEPARATOR " ") gradeAbbrvs
+        from Placement pl
+        where pl.dateFrom <= current_date() and pl.dateTo >= current_date()
+        group by pl.traineeId
```
[the remark this answers is not in the document]
Comment (CaiWillis) [reply tcs-persistence/src/main/resources/queries/traineeInfoForConnection.sql:57]: Will take a look. For the "Bulk Query", I'm getting 257828 rows returned in 20s

[the remark this answers is not in the document]
Comment (CaiWillis) [reply tcs-persistence/src/main/resources/queries/traineeInfoForConnection.sql:57]: Using the query as is but including the whereclauses I get a result in 11 seconds

[the remark this answers is not in the document]
Comment (CaiWillis) [reply tcs-persistence/src/main/resources/queries/traineeInfoForConnection.sql]: Have added an individual clause for placements and this reduces the time to < 1 second

File: tcs-service/src/main/java/com/transformuk/hee/tis/tcs/service/service/impl/RevalidationServiceImpl.java [hunk 2]
```diff
@@ -228,10 +228,12 @@ public ConnectionInfoDto buildTcsConnectionInfo(Long personId) {
 
     // $1 and $2 will be injected with the parenthesized match sub patterns from the RegEx
     final String whereClause = String.format("where $1.$2 = %s", personId);
+    final String individualPlacementClause = String.format("and pl.traineeId = %s", personId);
 
     final String query = sqlQuerySupplier
         .getQuery(SqlQuerySupplier.TRAINEE_CONNECTION_INFO)
         .replaceAll(WHERE_CLAUSE_MACRO_REGEX, whereClause)
+        .replace("INDIVIDUALPLACEMENTCLAUSE", individualPlacementClause)
```
Comment (JosephKelly) [inline tcs-service/src/main/java/com/transformuk/hee/tis/tcs/service/service/impl/RevalidationServiceImpl.java:236]: The WHERECLAUSE was modified and made more flexible to accommodate field names & table aliases.
If we are going to stick with a separate template for this clause, I'll ask that this uses that same method to reduce the likelihood of the table alias `pl` and field name changing in one place but not the other.

**Review by JosephKelly — CHANGES_REQUESTED** [review]
I think adding an extra clause like this makes it less consistent with the other persistence queries.
The proposed SQL does mean we have 1 less level of nesting but the consistency is better for me.

This commit is more consistent with the way that programme memberships are restricted: https://github.com/Health-Education-England/TIS-TCS/compare/main...feat/addCurrentPlacementGrade_TIS21-4769

I've left a couple of inline comments.

We also need to increment the version of the tcs-persistence module & use that in the `tcs-service` module.

File: tcs-persistence/src/main/resources/queries/traineeInfoForConnection.sql [hunk 3]
```diff
@@ -46,8 +47,19 @@ from (
...
+  		GROUP_CONCAT(pl.gradeAbbreviation SEPARATOR " ") gradeAbbrvs
+        from Placement pl
+        where pl.dateFrom <= current_date() and pl.dateTo >= current_date()
+        INDIVIDUALPLACEMENTCLAUSE
+        group by pl.traineeId
+        ) placements
+          on placements.traineeId = cd.id
   WHERECLAUSE(cd, id)
-) as ot
+) as ot where ot.gradeAbbrvs is null or ot.gradeAbbrvs not LIKE '%F1%' -- include trainees without current placements and exclude f1
```
[the remark this answers is not in the document]
Comment (CaiWillis) [reply tcs-persistence/src/main/resources/queries/traineeInfoForConnection.sql:62]: Would it work if I changed the group delimeter from a space to symbol (for argument's sake, | ) and did like %|F1|%?

File: tcs-service/src/main/java/com/transformuk/hee/tis/tcs/service/service/impl/RevalidationServiceImpl.java (hunk 2 above)
Comment (CaiWillis) [reply tcs-service/src/main/java/com/transformuk/hee/tis/tcs/service/service/impl/RevalidationServiceImpl.java:236]: Cool, I could change it to an ANDWHERE clause

File: tcs-persistence/src/main/resources/queries/traineeInfoForConnection.sql [hunk 4]
```diff
@@ -46,8 +47,19 @@ from (
     WHERECLAUSE(cm2, personId)
     group by cm2.programmeMembershipUuid
   ) latestCm on latestCm.programmeMembershipUuid = pm1.uuid
+  left join (
+      select
+  		pl.traineeId,
+  		-- one row per trainee
+  		GROUP_CONCAT(pl.gradeAbbreviation SEPARATOR " ") gradeAbbrvs
```
[the remark this answers is not in the document]
Comment (CaiWillis) [reply tcs-persistence/src/main/resources/queries/traineeInfoForConnection.sql:54]: 🤦 thanks!

**Comment by CaiWillis** [comment]
- Have removed the ANDWHERE clause and used the nested example
- Have added Regex and changes to group concat as suggested

Tested and the refactored query returns the same number of results as before, which is exatcly the number of F1 doctors less than the query before this commit, so all good

File: tcs-service/src/main/java/com/transformuk/hee/tis/tcs/service/service/impl/RevalidationServiceImpl.java (hunk 2 above)
Comment (CaiWillis) [reply tcs-service/src/main/java/com/transformuk/hee/tis/tcs/service/service/impl/RevalidationServiceImpl.java]: Have used the nested query as suggested

File: tcs-persistence/src/main/resources/queries/traineeInfoForConnection.sql [hunk 5]
```diff
@@ -46,8 +47,21 @@ from (
...
   ) latestCm on latestCm.programmeMembershipUuid = pm1.uuid
+  left join (
+      select
+  		pl.traineeId,
+  		-- one row per trainee
+  		GROUP_CONCAT(distinct pl.gradeAbbreviation SEPARATOR " | ") currentGrades
+      from (
+        select traineeId, gradeAbbreviation, dateFrom, dateTo
+        from Placement pl
+        WHERECLAUSE(pl, id)
```
Comment (hihilary) [inline tcs-persistence/src/main/resources/queries/traineeInfoForConnection.sql:58]: I guess it's traineeId you want to filter here.

**Comment by CaiWillis** [comment]
Note: change of approach, this PR is now intended to send the grade across to reval

File: tcs-persistence/src/main/resources/queries/traineeInfoForConnection.sql [hunk 6]
```diff
@@ -8,6 +8,7 @@ from (
     pm1.programmeMembershipType,
     pm1.programmeStartDate,
     pm1.programmeEndDate,
+    currentGrade.currentgrades,
```
Comment (hihilary) [inline tcs-persistence/src/main/resources/queries/traineeInfoForConnection.sql:11]: We need to use Camel Case too for the column name.

File: tcs-service/src/main/java/com/transformuk/hee/tis/tcs/service/service/impl/RevalidationServiceImpl.java [hunk 7]
```diff
@@ -66,14 +66,14 @@ public class RevalidationServiceImpl implements RevalidationService {
   private static final String PROGRAMME_NAME_FIELD = "programmeName";
   private static final String PROGRAMME_START_DATE_FIELD = "programmeStartDate";
   private static final String SURNAME_FIELD = "surname";
+  private static final String PLACEMENT_GRADE_FIELD = "currentgrades";
```
Comment (hihilary) [inline tcs-service/src/main/java/com/transformuk/hee/tis/tcs/service/service/impl/RevalidationServiceImpl.java:69]: The same here, camel case for column name.

**Review by hihilary — CHANGES_REQUESTED** [review]

**Review by hihilary — APPROVED** [review]

**Pull request merged**